File: sorting/non_overlapping_intervals/non_overlapping_intervals.py

def erase_overlapping_intervals(intervals: list[list[int]]) -> int:  # LeetCode Q.435.
    initial_size = len(intervals)
    intervals.sort()
    final_intervals: list[list[int]] = [intervals.pop(0)]
    for start, end in intervals:
        if final_intervals[-1][1] <= start:  # New interval can attach next to last interval.
            final_intervals.append([start, end])

        if end < final_intervals[-1][1]:
            # New interval is entirely inside last interval.
            final_intervals[-1] = [start, end]  # Replace last interval with new interval.

    return initial_size - len(final_intervals)

File: sorting/non_overlapping_intervals/non_overlapping_intervals.cpp
#include <vector>
using namespace std;

int erase_overlapping_intervals(vector<vector<int>> &intervals)
{ // LeetCode Q.435.
    // For a given interval start, always choose the smallest end.
    int initial_size = intervals.size();
    sort(intervals.begin(), intervals.end());

    vector<vector<int>> final_intervals = {intervals.front()};
    intervals.erase(intervals.begin()); // Base case.

    for (auto interval : intervals)
    {
        // Extend: last interval end <= current interval start.
        if (final_intervals.back()[1] <= interval[0])
        {
            final_intervals.push_back(interval);
        }
        // Replace: current interval end < last interval end.
        if (interval[1] < final_intervals.back()[1])
        {
            final_intervals.back() = interval;
        }
    }

    return initial_size - final_intervals.size();
}
